memcached/memcached_1.6.19.patch

```diff
@@ -0,0 +1,1308 @@
+                         break;
+                     }
+                     SSL_LOCK();
++#ifdef TLS
+                     ssl = SSL_new(settings.ssl_ctx);
++#elif WOLFSSL_MEMCACHED
++                    ssl = wolfSSL_new(settings.ssl_ctx);
++#endif
+                     SSL_UNLOCK();
+                     if (ssl == NULL) {
+                         if (settings.verbose) {
+@@ -3060,6 +3091,7 @@ static void drive_machine(conn *c) {
+                         close(sfd);
+                         break;
+                     }
++#ifdef TLS
+                     SSL_set_fd(ssl, sfd);
+                     int ret = SSL_accept(ssl);
+                     if (ret <= 0) {
+@@ -3076,6 +3108,45 @@ static void drive_machine(conn *c) {
+                             break;
+                         }
+                     }
++#elif WOLFSSL_MEMCACHED
++                    wolfSSL_set_fd(ssl, sfd);
++                    int ret = wolfSSL_accept(ssl);
++                    if (ret != WOLFSSL_SUCCESS) {
++                        int err = wolfSSL_get_error(ssl, ret);
++                        if (err != WOLFSSL_ERROR_WANT_READ && err != WOLFSSL_ERROR_WANT_WRITE) {
++                            if (settings.verbose) {
++                                fprintf(stderr, "SSL connection failed with error code : %d : %s\n",
++                                        err, wolfSSL_ERR_reason_error_string(err));
++                            }
++                            wolfSSL_free(ssl);
++                            close(sfd);
++                            STATS_LOCK();
++                            stats.ssl_handshake_errors++;
++                            STATS_UNLOCK();
++                            break;
++                        }
++                        else {
++                            /* Unlike OpenSSL, wolfSSL does not automatically retry
++                               on non-blocking connections, the application must
++                               repeat the call until it succeeds when WANT_READ/WANT_WRITE
++                               is returned. */
++                            int maxRetries = 10;
++                            do {
++                                struct pollfd to_poll[1];
++
++                                ret = wolfSSL_accept(ssl);
++                                err = wolfSSL_get_error(ssl, ret);
++                                if (err == WOLFSSL_ERROR_WANT_READ) {
++                                    to_poll[0].fd = sfd;
++                                    to_poll[0].events = POLLIN;
++                                    poll(to_poll, 1, 500);
++                                }
++                                maxRetries--;
++                            } while ((err == WOLFSSL_ERROR_WANT_READ || err == WOLFSSL_ERROR_WANT_WRITE)
++                                     && maxRetries > 0);
++                        }
++                    }
++#endif
```

Comment:
Does the `WOLFSSL_MODE_AUTO_RETRY_ATTEMPTS` define not take care of this? It looks like it has to be explicitly turned on but you can make it implicit for memcached.

Reply:
Note that this is a native wolfSSL port, auto retry is only included as part of our OpenSSL compatibility layer and did not appear to be trivial to break out.

Reply:
`ssl->ctx->autoRetry` is not gated on anything in the compatibility layer. The function `wolfSSL_CTX_set_mode` is very minimal. Just add the define to compile in `wolfSSL_CTX_set_mode` for memcached and enable it in memcached. Or if memcached doesn't disable this in any way then just enable it by default in `InitSSL_Ctx`.

Reply:
Will look into this, would certainly be cleaner than requiring memcached to do this.

Reply:
I was unfortunately only able to use this logic for wolfSSL_read, as that's the only function which it's implemented for in wolfSSL.  I still need memcached to handle retries for wolfSSL_write and wolfSSL_accept, or the tests break.
Still, this cleaned up the code a bit.

Reply:
Looking at the code for `WOLFSSL_MODE_AUTO_RETRY_ATTEMPTS` in `wolfSSLReceive`, it looks like it should be fairly easy to add the same functionality in `SendBuffered`. Have you looked into this?

Reply:
Done, now wolfSSL is handling auto retry for reads and writes